### PR TITLE
[WIP] 🌱 Enable webhook envtests

### LINKED ISSUE #5721: Enable envtest-based unit tests for webhooks
**User Story**

As a developer I would like to be able to write unit tests with envtest for our webhooks.

Note: In case of webhooks the fakeclient unit tests have (at least) the following limitations:
* The webhooks cannot use indices (as that's not supported by fakeclient)
* There is no marshalling/unmarshalling of Go types and no OpenAPI schema validation of the kube-apiserver as we directly call validate/default funcs with our structs. Especially with ClusterClass variables this can lead to cases where the unit test is successful but the apiserver would reject the create/update call without even asking the webhook

**Detailed Description**

I would suggest implementing it the same way we currently add indices to our testenv:
* create helper functions to register webhooks
* Add a `SetupWebhooks` func to `RunInput` (used in `internal/envtest.Run`)

This solution would enable envtest-based unit test for all our webhook packages except top-level `api/v1beta1`, because this would lead to a circular dependency between `api/v1beta` and `webhooks`. But we already have this limitation today and if we want to unit test one of the webhooks in the `api/v1beta1` package with envtest we can simply move it to the `webhooks` package.

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

WIP => currently mostly a PoC to discuss if we want to address the issue this way.

**DOES NOT WORK RIGHT NOW**

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5721

### CODE REVIEW

File: api/v1beta1/webhooks.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+package v1beta1
+
+import (
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// SetupWebhooksWithManager sets up all webhooks of the current package.
+func SetupWebhooksWithManager(mgr ctrl.Manager) error {
```
Comment (sbueringer) [inline api/v1beta1/webhooks.go:25]: We don't necessarily need the package level functions to avoid the circular dependencies, but they would avoid some repetition (otherwise we would have to inline those funcs, aka we would need the full list of resource in all those cases)

File: bootstrap/kubeadm/main.go [hunk 2]
```diff
@@ -216,18 +216,10 @@ func setupWebhooks(mgr ctrl.Manager) {
 		setupLog.Error(err, "unable to create webhook", "webhook", "KubeadmConfig")
 		os.Exit(1)
 	}
```
Comment (sbueringer) [inline bootstrap/kubeadm/main.go:218]: We don't have to register those webhooks (we don't register any other List webhooks)

File: internal/envtest/environment.go [hunk 3]
```diff
@@ -69,6 +68,9 @@ func init() {
 	klog.SetOutput(ginkgo.GinkgoWriter)
 
 	// Calculate the scheme.
+	// FIXME(tbd): if we want to enable envtests in those packages too we
```
Comment (sbueringer) [inline internal/envtest/environment.go:71]: TBD

File: internal/envtest/setup/webhook.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	kubeadmbootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	kcpv1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api/webhooks"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// AllWebhooksWithManager sets up all webhooks.
+func AllWebhooksWithManager(mgr ctrl.Manager) {
```
Comment (sbueringer) [inline internal/envtest/setup/webhook.go:34]: In theory we also don't need this helper to avoid circular dependencies, but the alternative would be to inline it and have the list of webhooks redundantly

File: internal/webhooks/suite_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"sigs.k8s.io/cluster-api/internal/envtest"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	env *envtest.Environment
+	ctx = ctrl.SetupSignalHandler()
+)
+
+// FIXME(sbueringer): just part of this PR to prove that it works.
```
Comment (sbueringer) [inline internal/webhooks/suite_test.go:34]: Just to prove the concept works, should be removed before merge

File: webhooks/webhooks.go [hunk 6]
```diff
@@ -14,14 +14,14 @@ See the License for the specific language governing permissions and
...
 import (
+	"sigs.k8s.io/cluster-api/internal/webhooks"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-func (r *KubeadmConfigList) SetupWebhookWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
-		Complete()
+// SetupWebhooksWithManager sets up all webhooks of the current package.
```
Comment (sbueringer) [inline webhooks/webhooks.go:24]: Note: that github diff is really weird (it's just a new file)

**Comment by sbueringer** [comment]
@killianmuldoon @fabriziopandini This PR is meant as a starting point for the discussion on how the solve the issue. 

tl;dr is
* we can't just register all webhooks in the envtest package because then envtest won't be usable in webhook envtests 
    * => so we have to move them out of there
* naive solution would be to just register all webhooks in the suite_test.go files.
    * => To avoid the redundancy I added two layers of helper, which both are not strictly necessary to avoid the circular dependency but avoid a lot of repetition.
        * Tests which require all webhooks can just use the setup func
        * webhook envtests have to individually register the required webhooks (because they cannot use the helper func because this would introduce a circular dependency)

P.S. apidiff would be fixed via deprecation if we want

**Comment by sbueringer** [comment]
I've discussed the issue with Fabrizio and we would simply implement the webhooks integration test in a package like `internal/webhooks/integration` instead of refactoring envtest like proposed in this PR.

The problem with this PR is that it distributes the webhook setup all over the place vs. currently webhooks will just work out of the webhooks when using envtest.
